src/Mvc/src/Microsoft.AspNetCore.Mvc.NewtonsoftJson/NewtonsoftJsonHelper.cs

```diff
@@ -60,54 +58,30 @@ public IHtmlContent Serialize(object value, JsonSerializerSettings serializerSet
                 throw new ArgumentNullException(nameof(serializerSettings));
             }
 
-            var jsonOutputFormatter = new NewtonsoftJsonOutputFormatter(serializerSettings, _charPool);
-
-            return SerializeInternal(jsonOutputFormatter, value);
+            var jsonSerializer = CreateHtmlSafeSerializer(serializerSettings);
+            return Serialize(value, jsonSerializer);
         }
 
-        private IHtmlContent SerializeInternal(NewtonsoftJsonOutputFormatter jsonOutputFormatter, object value)
+        private IHtmlContent Serialize(object value, JsonSerializer jsonSerializer)
         {
-            var stringWriter = new StringWriter(CultureInfo.InvariantCulture);
-            jsonOutputFormatter.WriteObject(stringWriter, value);
+            using (var stringWriter = new StringWriter(CultureInfo.InvariantCulture))
+            {
+                var jsonWriter = new JsonTextWriter(stringWriter)
+                {
+                    ArrayPool = _charPool,
+                };
 
-            return new HtmlString(stringWriter.ToString());
+                jsonSerializer.Serialize(jsonWriter, value);
```

Comment:
Wrap `jsonWriter` in a `using` that ends here. Otherwise, can't be sure that writer has been flushed or closed before creating the `HtmlString`.